sklearn/tree/_tree.pyx

```diff
@@ -1230,3 +1228,399 @@ cdef class Tree:
+        self.impurities = np.zeros(shape=(node_count), dtype=np.float64)
+        self.count = 0
+
+    cdef int save_metrics(self,
+                          DOUBLE_t effective_alpha,
+                          DOUBLE_t subtree_impurities) nogil:
+        self.ccp_alphas[self.count] = effective_alpha
+        self.impurities[self.count] = subtree_impurities
+        self.count += 1
+
+
+cdef _cost_complexity_prune(
+    Tree orig_tree,
+    _CCPPruneController controller,
+    unsigned char[:] leaves_in_subtree,
+    SIZE_t[:] capacity):
+    """Performs cost complexity pruning.
+
+    The ccp_controller signals when the pruning should stop.
+
+    Parameters
+    ----------
+    orig_tree : Tree
+        Original tree
+    ccp_controller : _CCPPruneController
+        cost complexity controller
+    leaves_in_subtree : unsigned char[:]
+        output for leaves of subtree
+    capacity : SIZE_T*
+        output for capacity of subtree
+    """
+
+    cdef:
+        SIZE_t i
+        SIZE_t n_nodes = orig_tree.node_count
+        # prior probability using weighted samples
+        DOUBLE_t[:] weighted_n_node_samples = orig_tree.weighted_n_node_samples
+        DOUBLE_t total_sum_weights = weighted_n_node_samples[0]
+        DOUBLE_t[:] impurity = orig_tree.impurity
+        # weighted impurity of each node
+        DOUBLE_t[:] r_node = np.empty(shape=n_nodes, dtype=np.float64)
+
+        SIZE_t[:] child_l = orig_tree.children_left
+        SIZE_t[:] child_r = orig_tree.children_right
+        SIZE_t[:] parent = np.zeros(shape=n_nodes, dtype=np.intp)
+
+        # Only uses the start and parent variables
+        Stack stack = Stack(INITIAL_STACK_SIZE)
+        StackRecord stack_record
+        int rc = 0
+        SIZE_t node_idx
+
+        SIZE_t[:] n_leaves = np.zeros(shape=n_nodes, dtype=np.intp)
+        DOUBLE_t[:] r_branch = np.zeros(shape=n_nodes, dtype=np.float64)
+        DOUBLE_t current_r
+        SIZE_t leaf_idx
+        SIZE_t parent_idx
+
+        # candidate nodes that can be pruned
+        unsigned char[:] candidate_nodes = np.zeros(
+            shape=n_nodes, dtype=np.uint8)
+        # nodes in subtree
+        unsigned char[:] in_subtree = np.ones(shape=n_nodes, dtype=np.uint8)
+        DOUBLE_t[:] g_node = np.zeros(shape=n_nodes, dtype=np.float64)
+        SIZE_t pruned_branch_node_idx
+        DOUBLE_t subtree_alpha
+        DOUBLE_t effective_alpha
+        SIZE_t child_l_idx
+        SIZE_t child_r_idx
+        SIZE_t n_pruned_leaves
+        DOUBLE_t r_diff
+        DOUBLE_t max_float64 = np.finfo(np.float64).max
+
+    # find parent node ids and leaves
+    with nogil:
+
+        for i in range(r_node.shape[0]):
+            r_node[i] = (
+                weighted_n_node_samples[i] * impurity[i] / total_sum_weights)
+
+        # Push root node, using StackRecord.start as node id
+        rc = stack.push(0, 0, 0, -1, 0, 0, 0)
+        if rc == -1:
+            with gil:
+                raise MemoryError()
```

Comment:
do we want to add a message like `Ran out of memory while pruning the tree` whenever you raise a `MemoryError` here?

Reply:
From the Python docs, a `MemoryError` is "Raised when an operation runs out of memory". Nevertheless I think adding a message about pruning would be useful. How about `MemoryError("pruning tree")`?

Reply:
Yep, that makes sense.

Reply:
I think you missed this one in your latest commit